OneCommunityGlobal/HighestGoodNetworkApp
# Tim add Materials List page

# Description
Add the Materials List page (Phase 2), a table of all purchased materials being used by the organization in their building projects.
- [Docs](https://docs.google.com/document/d/1WONWQW_56pXlyZpdQFibddFuT3iMxFjho-wfGwbHFV8/edit#bookmark=id.shlm4iwupd4s)
- [Wireframes](https://www.figma.com/file/kpwXZP2pLtOJ1Fn2u7Sivn/HGN-PII?node-id=505%3A5786)


## Related PRS (if any):
To test this PR you need to checkout [backend PR #567](https://github.com/OneCommunityGlobal/HGNRest/pull/567)

## Main changes explained:
- Adds route “bmdashboard/materials-list”
- Adds **bmdashboard** subdirectory to the actions, components, and reducers folders with related JS, JSX, and CSS files.
- On page load, an API request is sent to the backend to fetch the data (see backend PR for more). The response (either an error or the data) is added to Redux state and mapped to `props`.
- Based on the fetch result, the page will display either the error, the data, or a zero results page if there is no data.
   - Note: data values are arbitrary
- Data can be sorted by project
- Each data entry contains additional Usage, Updates, and Purchase info accessible via a button that opens a modal. There are also edit icons that currently return null but will eventually access Edit forms.

## How to test:
- Follow steps to test backend PR 567
- `git checkout Tim_bm_materials_list_index`
- `npm install`
- `npm run start:local`
- Log into HGN as a volunteer
- Verify your credentials at “localhost:3000/bmdashboard”, then navigate to “localhost:3000/bmdashboard/materials-list”
   - Note: you will need to manually enter this into your URL bar as there are currently no nav links
- The Materials List page should display a table with about 25 rows of dummy data by default
- Test the Projects filter properly filters the results by project
- Test the “View” buttons for various rows to test that a modal with new data is displayed
- Test modal overflow: in row Tim Test Project 1 > River Sand (default first row), open the Usage, Updates, and Records modals. Each have enough data rows added to ensure the modal will overflow the results. Try scrolling the modal to test overflow is working properly.
- Test the page and modals are adaptive to page width.
   - Modal widths are reactstrap defaults (I cannot figure out how to change this) and not very good. I added overflow styling so the modals should be able to handle x-axis overflow at narrower screen widths.
   - Main page styling is also mostly reactstrap defaults, but the table should also handle x-axis overflow at narrower screen widths.
- Test error state: if the fetch attempt returns an error, the page will display the error information with a button to reload the page and try again.
   - Easy way to test client-side error: open the file **actions/bmdashboard/materialsAction.js**. In the **fetchAllMaterials** function, edit the `ENDPOINTS.BM_MATERIALS_LIST` string so the request is sent to an incorrect endpoint (404 error).
      - Refresh the page to send a fresh request
   - Easy to way to test server-side error: deactivate the backend so that no response object is included in the error object (503 error)
- Test ‘no results’ state: if no data is returned from the database, table layout will be replaced with “No materials data” text and the Project filter will be disabled
   - How to test: in fetchAllMaterials func, change `dispatch(setMaterials(res.data))` to `dispatch(setMaterials([]))`
   - Note: you made need to refresh the page twice

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/98129080/bba52a8a-b180-4206-9d0e-18c65ee5377f

## Notes:
- The Edit icons (pencils) currently are not hooked up to anything and simply return null
- Error toasts that may pop up from a server-side error are coming from other error handling in the app
- Some reviewers are experiencing an issue where "/bmdashboard/materials-list" is displaying the error page instead of the data. To fix try running `git pull` then `npm run build` on the backend branch to make sure the branch is up to date.
